File: platforms/android/lib/src/main/java/com/shopify/checkoutkit/ComposedCheckoutCommunicationClient.kt
package com.shopify.checkoutkit

import kotlinx.serialization.SerializationException

/**
 * Composes a merchant-supplied protocol client with kit-owned default handlers.
 *
 * The default bindings make the dispatch policy explicit in one place:
 * request delegations such as [CheckoutProtocol.windowOpen] only fall back to the
 * kit default when the merchant does not return a response, while mandatory kit
 * notifications such as [CheckoutProtocol.error] always run after the merchant client.
 */
internal class ComposedCheckoutCommunicationClient(
    private val merchant: CheckoutCommunicationClient?,
    private val defaults: Map<String, DefaultClientBinding>,
) : CheckoutCommunicationClient {
    override fun process(message: String): String? {
        val method = method(message) ?: return merchant?.process(message)
        var response = merchant?.process(message)

        defaults[method]?.let { binding ->
            when (binding.policy) {
                DefaultClientPolicy.AlwaysRunAfterMerchant -> {
                    val defaultResponse = binding.client.process(message)
                    response = response ?: defaultResponse
                }
                DefaultClientPolicy.RunIfUnhandled ->
                    if (response == null) {
                        response = binding.client.process(message)
                    }
            }
        }

        return response
    }

    private fun method(message: String): String? = try {
        CheckoutProtocol.json.decodeFromString<EcpRequest>(message).method
    } catch (_: SerializationException) {
        null
    }
}

internal data class DefaultClientBinding(
    val client: CheckoutCommunicationClient,
    val policy: DefaultClientPolicy,
)

internal enum class DefaultClientPolicy {
    AlwaysRunAfterMerchant,
    RunIfUnhandled,
}

File: platforms/swift/Sources/ShopifyCheckoutKit/ComposedCheckoutCommunicationClient.swift
#if !COCOAPODS
    import EmbeddedCheckoutProtocol
#endif
import Foundation

/// Composes a merchant-supplied protocol client with kit-owned default handlers.
///
/// The default bindings make the dispatch policy explicit in one place:
/// requests such as `CheckoutProtocol.ready` and `CheckoutProtocol.windowOpen` fall
/// back to kit defaults when the merchant does not return a response, while mandatory
/// kit notifications such as `CheckoutProtocol.error` always run after the merchant
/// client.
struct ComposedCheckoutCommunicationClient: CheckoutCommunicationProtocol {
    let merchant: (any CheckoutCommunicationProtocol)?
    let defaults: [String: DefaultClientBinding]

    func process(_ message: String) async -> String? {
        guard let method = Self.method(message), let binding = defaults[method] else {
            return await merchant?.process(message)
        }

        switch binding.policy {
        case .kitOwned:
            return await binding.client.process(message)

        case .alwaysRunAfterMerchant:
            let response = await merchant?.process(message)
            let defaultResponse = await binding.client.process(message)
            return response ?? defaultResponse

        case .runIfUnhandled:
            if let response = await merchant?.process(message) {
                return response
            }
            return await binding.client.process(message)
        }
    }

    private static func method(_ message: String) -> String? {
        guard let request = try? JSONDecoder().decode(MethodEnvelope.self, from: Data(message.utf8)) else { return nil }
        return request.method
    }
}

private struct MethodEnvelope: Decodable {
    let method: String
}

struct DefaultClientBinding {
    let client: any CheckoutCommunicationProtocol
    let policy: DefaultClientPolicy
}

enum DefaultClientPolicy {
    case kitOwned
    case alwaysRunAfterMerchant
    case runIfUnhandled
}
